src/frontend/src/optimizer/plan_node/plan_base.rs

```diff
@@ -14,47 +14,132 @@
 
 use educe::Educe;
 use fixedbitset::FixedBitSet;
-use paste::paste;
 use risingwave_common::catalog::Schema;
 
 use super::generic::GenericPlanNode;
 use super::*;
-use crate::for_all_plan_nodes;
 use crate::optimizer::optimizer_context::OptimizerContextRef;
 use crate::optimizer::property::{Distribution, FunctionalDependencySet, Order};
 
-/// the common fields of all nodes, please make a field named `base` in
-/// every planNode and correctly value it when construct the planNode.
-#[derive(Clone, Debug, Educe)]
-#[educe(PartialEq, Eq, Hash)]
-pub struct PlanBase {
-    #[educe(PartialEq(ignore))]
-    #[educe(Hash(ignore))]
-    pub id: PlanNodeId,
-    #[educe(PartialEq(ignore))]
-    #[educe(Hash(ignore))]
-    pub ctx: OptimizerContextRef,
-    pub schema: Schema,
-    /// the pk indices of the PlanNode's output, a empty stream key vec means there is no stream key
-    pub stream_key: Option<Vec<usize>>,
-    /// The order property of the PlanNode's output, store an `&Order::any()` here will not affect
-    /// correctness, but insert unnecessary sort in plan
-    pub order: Order,
+/// Common extra fields for physical plan nodes.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+struct PhysicalCommonExtra {
     /// The distribution property of the PlanNode's output, store an `Distribution::any()` here
     /// will not affect correctness, but insert unnecessary exchange in plan
-    pub dist: Distribution,
+    dist: Distribution,
+}
+
+/// Extra fields for stream plan nodes.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+struct StreamExtra {
+    /// Common fields for physical plan nodes.
+    physical: PhysicalCommonExtra,
+
     /// The append-only property of the PlanNode's output is a stream-only property. Append-only
     /// means the stream contains only insert operation.
-    pub append_only: bool,
+    append_only: bool,
     /// Whether the output is emitted on window close.
-    pub emit_on_window_close: bool,
-    pub functional_dependency: FunctionalDependencySet,
+    emit_on_window_close: bool,
     /// The watermark column indices of the PlanNode's output. There could be watermark output from
     /// this stream operator.
-    pub watermark_columns: FixedBitSet,
+    watermark_columns: FixedBitSet,
+}
+
+/// Extra fields for batch plan nodes.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+struct BatchExtra {
+    /// Common fields for physical plan nodes.
+    physical: PhysicalCommonExtra,
+
+    /// The order property of the PlanNode's output, store an `&Order::any()` here will not affect
+    /// correctness, but insert unnecessary sort in plan
+    order: Order,
```

Comment:
Putting order here is a little counter-intuitive, since our materialized view could have an order by property i.e. `StreamMaterialize` theoretically could access its `order` property, but now it would cause a panic. However, I don't have a better idea right now.

Reply:
> `StreamMaterialize` theoretically could access its `order` property, but now it would cause a panic.

Even though we now force a panic when accessing a bad field, all tests are passing. I'm also surprised that this doesn't seem to be a problem.